### PR TITLE
docs: add tutorial for writing Lua plugin

### PR DESCRIPTION
**Pull request by echasnovski**
Problem: there is no built-in tutorial on how to write Lua plugin from
  scratch.

Solution: write the tutorial.

------

This is a (currently draft) suggestion for built-in tutorial on how to write Lua plugin. The general idea is to list some basic practices in an encouraging tone accompanied by small digestible examples and lots of links.

Most of necessary Lua Neovim background is already there in a very informative 'lua-guide.txt'. So I think that making a guide for writing a Lua plugin in the form of hands-on tutorial while *suggesting* some common approaches provides a nice balance to 'lua-guide.txt'.

Although a draft with only some sections written, I'd like to ask the whole @neovim/core for the feedback on whether this is worth pursuing along the lines outlined near the TODOs.

### CODE REVIEW

**Comment by justinmk** [comment]
This looks pretty different from https://github.com/nvim-neorocks/nvim-best-practices . I intentionally requested that https://github.com/nvim-neorocks/nvim-best-practices be upstreamed because it's a collection of easy to follow advice that is reasonable. I don't want to have a big discussion or debate about alternatives. We should start with what I outlined in https://github.com/nvim-neorocks/nvim-best-practices/issues/5 . Not something else.

File: runtime/doc/lua-plugin.txt [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+                    Tutorial on how to write a Lua plugin
+
+
+                                       Type |gO| to see the table of contents.
+
+==============================================================================
+Introduction                                                        *lua-plugin*
+
+Writing a Lua plugin from scratch can be overwhelming. This tutorial is aimed
```
Comment (justinmk) [inline runtime/doc/lua-plugin.txt:13]: this is all just way too verbose and conversational.

File: runtime/doc/lua-plugin.txt [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+The core functionality of any Lua plugin is its code. It is located in `lua/`
+subdirectory of the plugin.
+
+First start by creating a directory `no-trailing/` inside your 'runtimepath'.
+In this tutorial let's use the following path:
+`$XDG_DATA_HOME/nvim/site/pack/tutorial/start/no-trailing` (see |$XDG_DATA_HOME
+for its value on your system). The reason to choose this complicated path is
+because it utilizes built-in |packages| functionality to make this plugin
+installed in your config.
```
Comment (dundargoc) [inline runtime/doc/lua-plugin.txt:42]: The builtin packages are super sus, I wouldn't mention them in any guide.

Comment (dundargoc) [inline runtime/doc/lua-plugin.txt:42]: This feels overkill. You're already specifying later that `no-trailing` should be put in runtimepath.

File: runtime/doc/lua-plugin.txt [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+  trailing lines.
+
+• Any of `require('no-trailing').trim()`, `require('no-trailing').setup('a')`,
+  `require('no-trailing').trim(0, { empty = 'a' })` will throw an error.
+
+Notes:
+
+• Choosing function scope (i.e. what it does) is up to you. Usually preferring
+  the functionality just enough for it to be useful for target audience is
+  a good approach. You can see what Nvim does in |dev-lua|.
```
Comment (dundargoc) [inline runtime/doc/lua-plugin.txt:106]: Huh? This didn't help at all. I don't know what the takeaway from this is or what "guiding" is done here.

File: runtime/doc/lua-plugin.txt [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+• Choosing names for plugin/function/arguments/options is up to you. Usually
+  choosing a short "to the point" yet descriptive name is a good approach.
+  You can see what Nvim does in |dev-naming|.
+
+• Choosing function signature (i.e its arguments) is up to you. Preferring
+  first mandatory arguments followed by a single optional table `opts` with
+  optional arguments is a common convention.
+  You can see what Nvim uses in |dev-patterns| and |dev-api|.
```
Comment (dundargoc) [inline runtime/doc/lua-plugin.txt:115]: There's lotta "up to you"s here. Obviously people are free to go with their gut, don't think specifying this is necessary.

File: runtime/doc/lua-plugin.txt [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+• To interactively test latest plugin functionality, open separate fresh Nvim
+  instance with plugin loaded and interact with it.
```
Comment (dundargoc) [inline runtime/doc/lua-plugin.txt:118]: This isn't that helpful, ngl. I'd just yeet it.

File: runtime/doc/lua-plugin.txt [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+    -- Create user facing function
+    M.trim = function(buf_id, opts)
+      -- Validate required arguments
+      local ok, is_buf_valid = pcall(vim.api.nvim_buf_is_valid, buf_id)
+      assert(ok and is_buf_valid, '`buf_id` is not a valid buffer identifier')
+
+      -- Infer and validate optional arguments
+      opts = vim.tbl_deep_extend('force', { empty = false }, opts or {})
+      assert(type(opts.empty) == 'boolean', '`opts.empty` is not boolean')
```
Comment (dundargoc) [inline runtime/doc/lua-plugin.txt:66]: I'd honestly yeet all validation and error checking for the time being, I usually find it annoying when guides go "best practices" on me when I just want the thing to work first.

File: runtime/doc/lua-plugin.txt [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+      end
+      return 0
+    end
+
+    return M
+<
+
+Now there are two main approaches on how this can be used in main module:
+eager loading and lazy loading.
+
```
Comment (dundargoc) [inline runtime/doc/lua-plugin.txt:152]: There's a LOT of code to wade through here even though the relevant part is only like 3 lines. I wonder if the "build plugins from scratch" is the way to go here, it leads to *lot* of code reading.

Comment (justinmk) [reply runtime/doc/lua-plugin.txt:152]: +100. See also https://github.com/neovim/neovim/issues/22366#issuecomment-1445501976

File: runtime/doc/lua-plugin.txt [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+• Eager loading results in a more structured code with less lines.
+
+• Lazy loading results in a constant calls of `require()` during each function
+  execution. Although it caches its results, the time it spends to search the
+  cache is not strictly zero, which leads to a very small overhead (usually
+  order of fractions of a microsecond). This might matter for frequently
+  called functions.
+
+Usually both approaches are fine. Using eager loading and preferring lazy
+loading for expensive to source modules is usually a good idea.
```
Comment (dundargoc) [inline runtime/doc/lua-plugin.txt:216]: Like, I don't think this is wrong but this is just too much. I just want something quick and easy to get started, not a deepdive into everything I might possible use. If both approaches are *usually* fine then we could just pick one and call it a day. More information is not better.

**Review by dundargoc — CHANGES_REQUESTED** [review]
I think there is a mismatch in the expectation of who the target audience is, or perhaps what the purpose of the guide is. In my mind the purpose of such a guide is to get up and running fast; a starting point that can be improved upon with time. I *expect* (and appreciate) some parts being deliberately left out for the time being. The last thing I'd personally want is to ponder on whether my design requires lazy or eager loading when writing my first plugin.

**Pull request closed without merging**